### ISSUE TITLE
PyTorch Functionality

### ISSUE THREAD

**Issue by glenn-jocher** [ialhashim/DenseDepth#19]
Hello, thank you for your implementation! I have a few questions regarding PyTorch:

- Does your PyTorch port currently work correctly? 
- Are there pretrained weights available in PyTorch format?
- How do the PyTorch test results compare to Keras?
- Is inference possible on CPU? How long would this take on a small image (i.e. 640x640)?

**Comment by ialhashim** [comment]
- The PyTorch port unfortunately does not produce the same performance as the Keras one. One major thing might be in the loss function. Unfortunately, PyTroch does not provide a standard implementation of the SSIM function. My own experiments training with PyTroch visually result in worse models. I hope someone figures out why.
- Unfortunately not, but if you train on a single card that can handle `batch size = 8` it should take like less than a day to get some reasonable results on the NYU data set. This is true for both PyTorch and Keras. You should see some learning in the first epoch.
- Keras implementation beats all existing state of the art, PyTorch is worse than state of the art.
- Yes, I am not sure how long but it won't be real time. Possibly less than a minute for each image.

**Comment by leonzfa** [comment]
Maybe there is something wrong with DepthNorm function?
```
def DepthNorm(depth, maxDepth=1000.0): 
    return maxDepth / depth
```
It should be "depth / maxDepth"?

**Comment by tonysy** [comment]
@ialhashim  Would you like to share the performance produced with the pytorch code on NYU dataset. I want to the gap is how much?  I plan to tune the code and try to improve the performance on pytorch code. Thanks.

**Comment by anshulrai** [comment]
> * The PyTorch port unfortunately does not produce the same performance as the Keras one. One major thing might be in the loss function. Unfortunately, PyTroch does not provide a standard implementation of the SSIM function. My own experiments training with PyTroch visually result in worse models. I hope someone figures out why.


@ialhashim Is it because you haven't incorporated the `l_edges` loss term (the L1 loss defined over the image gradient of the depth image) which is there in the Keras loss function? 
Currently on line 70 of the PyTorch train.py file you have only two terms:
 `loss = (1.0 * l_ssim) + (0.1 * l_depth)`
While in your Keras loss function you have three terms:
`(w1 * l_ssim) + (w2 * K.mean(l_edges)) + (w3 * K.mean(l_depth))`

**Comment by ialhashim** [comment]
> @ialhashim Would you like to share the performance produced with the pytorch code on NYU dataset. I want to the gap is how much? I plan to tune the code and try to improve the performance on pytorch code. Thanks.
As far as I remember it was placing our network below state of the art so that was a no go for me. Unfortunately I don't have a saved trained PyTorch model. One thing to try is to evaluate the metrics for say 5 or 10 epochs on both implementations (make sure to disable or enable augmentation on both implementation) then see where they stand. I don't think you need to train very long before you see there is a gap.

**Comment by ialhashim** [comment]
> @ialhashim Is it because you haven't incorporated the `l_edges` loss term (the L1 loss defined over the image gradient of the depth image) which is there in the Keras loss function?
> Currently on line 70 of the PyTorch train.py file you have only two terms:
> `loss = (1.0 * l_ssim) + (0.1 * l_depth)`
> While in your Keras loss function you have three terms:
> `(w1 * l_ssim) + (w2 * K.mean(l_edges)) + (w3 * K.mean(l_depth))`

In my earlier PyTorch implementation I did include the same loss in both. I suspect two issues that might contribute to the performance discrepancy that are hard to debug: (1) the SSIM implementation in PyTorch is not as good as the ones provided by TF (2) Some default parameters used for the different layers (I don't know which) on Keras are set to a good value where as in PyTorch it expects the user to set them. But I am just speculating here, it could be some other issue.

**Comment by tonysy** [comment]
@ialhashim  Thanks for your reply. Would you like to share the evaluation code to compute the metric(rmse, lg and etc.) for PyTorch version? Thanks.

**Comment by ialhashim** [comment]
It shouldn't be that much different from the one I used for Keras. You simply need the ground truth and a predicted depth image and use the same function I am using in the evaluation script.

**Comment by ugamman** [comment]
> * The PyTorch port unfortunately does not produce the same performance as the Keras one. One major thing might be in the loss function. Unfortunately, PyTroch does not provide a standard implementation of the SSIM function. My own experiments training with PyTroch visually result in worse models. I hope someone figures out why.
> * Unfortunately not, but if you train on a single card that can handle `batch size = 8` it should take like less than a day to get some reasonable results on the NYU data set. This is true for both PyTorch and Keras. You should see some learning in the first epoch.
> * Keras implementation beats all existing state of the art, PyTorch is worse than state of the art.
> * Yes, I am not sure how long but it won't be real time. Possibly less than a minute for each image.

Hi,  it's now available SSIM functionn in pytorch: https://pypi.org/project/pytorch-msssim/
